espnet2/asr_transducer/error_calculator.py

```diff
@@ -0,0 +1,168 @@
+"""Error Calculator module for Transducer."""
+
+from typing import List, Tuple
+
+import torch
+
+from espnet2.asr_transducer.beam_search_transducer import BeamSearchTransducer
+from espnet2.asr_transducer.decoder.abs_decoder import AbsDecoder
+from espnet2.asr_transducer.joint_network import JointNetwork
+
+
+class ErrorCalculator(object):
+    """Calculate CER and WER for transducer models.
+
+    Args:
+        decoder: Decoder module.
+        joint_network: Joint Network module.
+        token_list: List of token units.
+        sym_space: Space symbol.
+        sym_blank: Blank symbol.
+        report_cer: Whether to compute CER.
+        report_wer: Whether to compute WER.
+
+    """
+
+    def __init__(
+        self,
+        decoder: AbsDecoder,
+        joint_network: JointNetwork,
+        token_list: List[int],
+        sym_space: str,
+        sym_blank: str,
+        report_cer: bool = False,
+        report_wer: bool = False,
+    ):
+        """Construct an ErrorCalculatorTransducer."""
+        super().__init__()
+
+        self.beam_search = BeamSearchTransducer(
+            decoder=decoder,
+            joint_network=joint_network,
+            beam_size=1,
+            search_type="default",
+            score_norm=False,
+        )
+
+        self.decoder = decoder
+
+        self.token_list = token_list
+        self.space = sym_space
+        self.blank = sym_blank
+
+        self.report_cer = report_cer
+        self.report_wer = report_wer
+
+    def __call__(self, encoder_out: torch.Tensor, target: torch.Tensor):
+        """Calculate sentence-level WER/CER score for Transducer model.
+
+        Args:
+            encoder_out: Encoder output sequences. (B, T, D_enc)
+            target: Target label ID sequences. (B, L)
+
+        Returns:
+            : Sentence-level CER score.
+            : Sentence-level WER score.
+
+        """
+        cer, wer = None, None
+
+        batchsize = int(encoder_out.size(0))
+
+        encoder_out = encoder_out.to(next(self.decoder.parameters()).device)
+
+        batch_nbest = [self.beam_search(encoder_out[b]) for b in range(batchsize)]
```

Comment:
Sorry, I don't fully understand here. Do we need beam search to calculate Acc/CER during the training?
Besides, it is possible to combine this class w/ `espnet/net2/e2e_asr_common.py:ErrorCaculator()`.

Reply:
> Sorry, I don't fully understand here. Do we need beam search to calculate Acc/CER during the training?

If CER/WER report is enabled, yes. If you're asking why we-re using beam search instead of greedy search, it's because of the unrestricted expansion. It's better for monitoring IMO.

Reply:
I see, thanks for the explanation.